documentation/design-docs/eventcounters.md

```diff
@@ -0,0 +1,306 @@
+        }
+    }
+
+    // Method being called from request handlers to log that a request happened
+    public void Request()
+    {
+        requestCnt += 1;
+    }
+```
+
+`RequestEventSource.Request()` can be called from a request handler, and `requestRateCounter` simply polls this value at the interval specified by the consumer of this counter. However, this method can be called by multiple threads at once, so `requestCnt` can be susceptible to race.
+
+Therefore, this method should be modified to the following to prevent the race.
+```cs
+public void Request()
+{
+    Interlocked.Increment(ref _requestCnt);
+}
+```
+
+## Consuming EventCounters
+
+There are two main ways of consuming EventCounters: in-proc and out-of-proc. 
+
+### Consuming in-proc
+
+You can consume the counter values via the `EventListener` API. `EventListener` is an in-proc way of consuming any Events written by all instances of EventSources in your application. For more details on how to use the EventListener API, refer to the [EventListener documentation](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventlistener).
+
+First, the EventSource that produces the counter value needs to be enabled. To do this, you can override the `OnEventSourceCreated` method to get a notification when an EventSource is created, and if this is the correct EventSource with your EventCounters, then you can call Enable on it. Here is an example of such override:
+
+```cs
+protected override void OnEventSourceCreated(EventSource source)
+{
+    if (source.Name.Equals("System.Runtime"))
+    {
+        Dictionary<string, string> refreshInterval = new Dictionary<string, string>()
+        {
+            { "EventCounterIntervalSec", "1" }
+        };
+        EnableEvents(source, 1, 1, refreshInterval);
+    }
+}
+```
+
+#### Sample Code
+
+This is a sample `EventListener` class that simply prints out all the counter names and values from a the .NET runtime's EventSource for publishing its internal counters (`System.Runtime`) at some interval.
+
+```cs
+public class SimpleEventListener : EventListener
+{        
+    private readonly EventLevel _level = EventLevel.Verbose;
+
+    public int EventCount { get; private set; } = 0;
+
+    private int _intervalSec;
+
+    public SimpleEventListener(int intervalSec)
+    {
+        _intervalSec = intervalSec;
+    }
+
+
+    protected override void OnEventSourceCreated(EventSource source)
+    {
+        if (source.Name.Equals("System.Runtime"))
+        {
+            Dictionary<string, string> refreshInterval = new Dictionary<string, string>() 
+            {
+                { "EventCounterIntervalSec", _intervalSec.ToString() }
+            };
+            EnableEvents(source, _level, (EventKeywords)(-1), refreshInterval);
+        }
+    }
+
+    private (string Name, string Value) getRelevantMetric(IDictionary<string, object> eventPayload)
+    {
+        string counterName = "";
```

Comment:
Since this simple example doesn't distinguish in the output between Mean and Increment, there is no need for the implementation to do so. We could have a single counterValue field which is set from either the Mean or the Increment.